libbfddp/bfddp_extra.c

```diff
@@ -0,0 +1,858 @@
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include <sys/time.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "bfddp_extra.h"
+
+/*
+ * Callback handling.
+ */
+
+static int
+bfddp_session_new_dummy(__attribute__((unused)) struct bfd_session *bs,
+			__attribute__((unused)) void *arg)
+{
+	return 0;
+}
+
+static void
+bfddp_session_update_dummy(__attribute__((unused)) struct bfd_session *bs,
+			   __attribute__((unused)) void *arg)
+{
+}
+
+static void
+bfddp_session_state_change_dummy(
+	__attribute__((unused)) struct bfd_session *bs,
+	__attribute__((unused)) void *arg,
+	__attribute__((unused)) enum bfd_state_value ostate,
+	__attribute__((unused)) enum bfd_state_value nstate)
+{
+}
+
+/* Our selected BFD integration callbacks. */
+struct bfddp_callbacks bfddp_callbacks;
+
+void
+bfddp_initialize(struct bfddp_callbacks *bc)
+{
+	bfddp_callbacks = *bc;
+
+	if (bfddp_callbacks.bc_session_new == NULL)
+		bfddp_callbacks.bc_session_new = bfddp_session_new_dummy;
+	if (bfddp_callbacks.bc_session_update == NULL)
+		bfddp_callbacks.bc_session_update = bfddp_session_update_dummy;
+	if (bfddp_callbacks.bc_session_free == NULL)
+		bfddp_callbacks.bc_session_free = bfddp_session_update_dummy;
+	if (bfddp_callbacks.bc_state_change == NULL)
+		bfddp_callbacks.bc_state_change =
+			bfddp_session_state_change_dummy;
+
+#define CALLBACK_CHECK(cb)                                                     \
+	do {                                                                   \
+		if (bfddp_callbacks.cb == NULL) {                              \
+			fprintf(stderr, "%s: callback " #cb " not set\n",      \
+				__func__);                                     \
+			exit(1);                                               \
+		}                                                              \
+	} while (0)
+
+	CALLBACK_CHECK(bc_tx_control);
+	CALLBACK_CHECK(bc_tx_control_update);
+	CALLBACK_CHECK(bc_tx_control_stop);
+	CALLBACK_CHECK(bc_rx_control_update);
+	CALLBACK_CHECK(bc_rx_control_stop);
+
+#undef CALLBACK_CHECK
```

Comment:
Remove `#undef`
Together with the previous comment. Define a prefix (e.g. `BFDDP_CALLBACK_CHECK`). `undef` is unnecessary as `.c` will not be included (hopefully) in other translation units.

Reply:
Makes sense, saves us one line of code.